### PR TITLE
Improve the efficiency by replace push_back with emplace_back

### PR DESCRIPTION
**Pull request by bingoko**
Improve the Dogecoin core efficiency without any major change.

### CODE REVIEW

**Comment by Cyclic** [comment]
> Improve the Dogecoin core efficiency without any major change.

“Copying of the constructor doesn’t take place while using the emplace_back() function. The cost due to the copying of elements is high.”

**Review by huy-springlabs — DISMISSED** [review]

**Review by Tracachang — DISMISSED** [review]

**Comment by rnicoll** [comment]
Very cautious that the semantics of the two are different, and while typically yes the one which doesn't copy the object will be faster, we'd need to verify the original object isn't then reused/modified elsewhere in ways which would cause problems.

**Comment by l4p4** [comment]
As complement of @rnicoll, be careful with this change; For more detail: https://stackoverflow.com/questions/4303513/push-back-vs-emplace-back

It is necessary to carry out tests, verifying all calls in the code. But I don't see that it brings any significant performance, it just reduces the possibilities in the long run and can cause issues

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2106,7 +2106,7 @@ void CWallet::AvailableCoins(vector<COutput> &vCoins, bool fOnlyConfirmed, const
                 bool fSpendableIn = ((mine & ISMINE_SPENDABLE) != ISMINE_NO) || (coinControl && coinControl->fAllowWatchOnly && (mine & ISMINE_WATCH_SOLVABLE) != ISMINE_NO);
                 bool fSolvableIn = (mine & (ISMINE_SPENDABLE | ISMINE_WATCH_SOLVABLE)) != ISMINE_NO;
 
-                vCoins.push_back(COutput(pcoin, i, nDepth, fSpendableIn, fSolvableIn));
+                vCoins.emplace_back(COutput(pcoin, i, nDepth, fSpendableIn, fSolvableIn));
```
Comment (nooperation) [inline src/wallet/wallet.cpp:2109]: `vCoins.emplace_back(COutput(pcoin, i, nDepth, fSpendableIn, fSolvableIn));`

This is a good place to show why you can't just simply find and replace all instances of `push_back` with `emplace back`. This is still allocating an entire temporary COutput object **A** to pass to `emplace_back`. Another COutput object is allocated for vCoins itself, moving/copying the contents of the temporary **A** object to itself. This is pretty much the same thing that push_back is doing so this solves nothing

However, consider the following instead:
`
vCoins.emplace_back(pcoin, i, nDepth, fSpendableIn, fSolvableIn);
`

Notice how even visually you're not allocating a temporary object to pass to `emplace_back`? This is where `emplace_back` begins to shine, but you will have to handle these changes on a case by case basis to make sure you're actually reducing allocations and not just adding more work

Here's a snippet of code you can play around with to see it in action: http://cpp.sh/2vzu5

**Comment by rnicoll** [comment]
Going to close and re-open to re-run CI.

**Review by rnicoll — CHANGES_REQUESTED** [review]
Appears to not compile. Needs to pass all tests on Travis before we'd consider it, and honestly very cautious this introduces risk for undefined reward.

**Comment by bingoko** [comment]
> Appears to not compile. Needs to pass all tests on Travis before we'd consider it, and honestly very cautious this introduces risk for undefined reward.

The problem is about couldn't find the emplace_back func in std::vector in the Travis server. It could be the version of C++ or OS system that we used are different.

**Comment by patricklodder** [comment]
> The problem is about couldn't find the emplace_back func in std::vector in the Travis server. It could be the version of C++ or OS system that we used are different.

Travis builds from `depends/`, as will be done on the gitian builds we need to build auditable binaries from. So it is very important that Travis passes.

**Comment by jimmyt1988** [comment]
Perhaps you could provide data of before and after. Perhaps at n=5 on all test cases you can think of that prove your performance gains would provide sufficient proof of gains? :). Lab coat anyone?

**Comment by patricklodder** [comment]
I'm re-configuring this PR to go against `1.14.7-dev`. I've now re-set the base 3 times, which is fine, however, for this to become mergable, it needs to pass tests. There will also be merge conflicts when this gets rebased (which it must.)

Perhaps an easier approach that doesn't create a lot of merge conflicts would be to:

1. create a benchmark in `src/bench` that executes a function where we think this would be beneficial for performance
2. separately modify the function to use `emplace_back` on top, so that we have 2 metrics.

@bingoko would that be something you'd be willing to pick up?

**Pull request closed without merging**